share/tasks.py

```diff
@@ -1,409 +1,187 @@
+                'log_id': log.id,
+                'source': log.source_config.source.long_title,
+                'source_config': log.source_config.label,
+            })
 
+            log.task_id = self.request.id
+            HarvestLog.objects.filter(id=log.id).update(task_id=self.request.id)
 
-class DisambiguatorTask(LoggedTask):
+        if log.completions > 0 and log.status == HarvestLog.STATUS.succeeded and not superfluous:
+            log.skip(HarvestLog.SkipReasons.duplicated)
+            logger.warning('%r has already been harvested. Force a re-run with superfluous=True', log)
+            return None
+        elif log.completions > 0 and log.status == HarvestLog.STATUS.succeeded:
+            logger.info('%r has already been harvested. Re-running superfluously', log)
 
-    def setup(self, normalized_id, *args, **kwargs):
-        self.normalized = NormalizedData.objects.get(pk=normalized_id)
-        self.source = self.normalized.source
+        if exhaust and log_id is None:
+            if force:
+                logger.warning('propagating force=True until queue exhaustion')
 
-    def do_run(self, *args, **kwargs):
-        # Load all relevant ContentTypes in a single query
-        ContentType.objects.get_for_models(*apps.get_models('share'), for_concrete_models=False)
+            logger.debug('Spawning another harvest task')
+            res = harvest.apply_async(self.request.args, self.request.kwargs)
+            logger.info('Spawned %r', res)
 
-        logger.info('%s started disambiguation for NormalizedData %s at %s', self.started_by, self.normalized.id, datetime.datetime.utcnow().isoformat())
+        logger.info('Harvesting %r', log)
 
         try:
-            with transaction.atomic():
-                cg = ChangeGraph(self.normalized.data['@graph'], namespace=self.normalized.source.username)
-                cg.process()
-                cs = ChangeSet.objects.from_graph(cg, self.normalized.id)
-                if cs and (self.source.is_robot or self.source.is_trusted or Source.objects.filter(user=self.source).exists()):
-                    # TODO: verify change set is not overwriting user created object
-                    cs.accept()
-        except Exception as e:
-            logger.exception('Failed disambiguation for NormalizedData %s. Retrying...', self.normalized.id)
-            raise self.retry(countdown=10, exc=e)
-
-        logger.info('Finished disambiguation for NormalizedData %s by %s at %s', self.normalized.id, self.started_by, datetime.datetime.utcnow().isoformat())
-
-
-class BotTask(AppTask):
-
-    def do_run(self, last_run=None, **kwargs):
-        bot = self.config.get_bot(self.started_by, last_run=last_run, **kwargs)
-        logger.info('Running bot %s. Started by %s', bot, self.started_by)
-        bot.run()
+            for datum in log.source_config.get_harvester().harvest_from_log(log, force=force, ignore_disabled=ignore_disabled, limit=limit):
```

Comment:
If `get_harvester` fails here (say, the source config uses a harvester from another branch), the harvest log isn't marked failed, and `harvest` keeps trying it over and over. I think either `harvest_from_log` should be a method on `HarvestLog` (so `get_harvester` is inside the `handle` block), or there should be better error handling in `harvest`.

Errors about twice a second:
```
[2017-05-26 13:27:45,964: WARNING/PoolWorker-4] Could not load io.tind
[2017-05-26 13:27:46,025: ERROR/PoolWorker-4] Task share.tasks.harvest[11cefce2-2e6a-4e8e-a72b-2e3af1ad7553] raised unexpected: NoMatches("No 'share.harvesters' driver found, looking for 'io.tind'",)
Traceback (most recent call last):
  File "/Users/abram/.virtualenvs/share/lib/python3.5/site-packages/celery/app/trace.py", line 367, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/Users/abram/.virtualenvs/share/lib/python3.5/site-packages/celery/app/trace.py", line 622, in __protected_call__
    return self.run(*args, **kwargs)
  File "/Users/abram/code/SHARE/share/tasks.py", line 162, in harvest
    countdown=(random.random() + 1) * min(settings.CELERY_RETRY_BACKOFF_BASE ** self.request.retries, 60 * 15)
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/abram/code/SHARE/share/models/logs.py", line 357, in lock_first
    yield item
  File "/Users/abram/code/SHARE/share/tasks.py", line 150, in harvest
    for datum in log.source_config.get_harvester().harvest_from_log(log, force=force, ignore_disabled=ignore_disabled, limit=limit):
  File "/Users/abram/code/SHARE/share/models/ingest.py", line 129, in get_harvester
    return self.harvester.get_class()(self, **{**kwargs, **(self.harvester_kwargs or {})})
  File "/Users/abram/code/SHARE/share/models/ingest.py", line 222, in get_class
    return driver.DriverManager('share.harvesters', self.key).driver
  File "/Users/abram/.virtualenvs/share/lib/python3.5/site-packages/stevedore/driver.py", line 61, in __init__
    warn_on_missing_entrypoint=warn_on_missing_entrypoint
  File "/Users/abram/.virtualenvs/share/lib/python3.5/site-packages/stevedore/named.py", line 89, in __init__
    self._init_plugins(extensions)
  File "/Users/abram/.virtualenvs/share/lib/python3.5/site-packages/stevedore/driver.py", line 113, in _init_plugins
    (self.namespace, name))
stevedore.exception.NoMatches: No 'share.harvesters' driver found, looking for 'io.tind'
```